# **Computational Drug Discovery [Part 1] Download Bioactivity Data**

Alberto Moa

---

# 1. Introduction

In this Jupyter notebook, I will be building a machine learning model using the ChEMBL bioactivity data.

**ChEMBL Database**

The [*ChEMBL Database*](https://www.ebi.ac.uk/chembl/) is a database that contains curated bioactivity data of more than 2 million compounds. It is compiled from more than 76,000 documents, 1.2 million assays and the data spans 13,000 targets and 1,800 cells and 33,000 indications.
[Data as of March 25, 2020; ChEMBL version 26].

# 2. Initialization
First, installing and importing useful libraries

## 2.1 Installing Libraries

Install the ChEMBL web service package to retrieve bioactivity data from the ChEMBL Database.

In [1]:
pip install chembl_webresource_client


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install pandas


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


## 2.2 Importing Libraries

In [3]:
# import necessary libraries
import pandas as pd
from chembl_webresource_client.new_client import new_client

## 2.3 Search for Target Protein
### 2.3.1 Target search for Coronavirus

In [4]:
# target search for coronavirus
target = new_client.target
target_query = target.search('coronavirus')
targets = pd.DataFrame.from_dict(target_query)
targets

cross_references  \
0                                                 []   
1                                                 []   
2                                                 []   
3                                                 []   
4                                                 []   
5                                                 []   
6  [{'xref_id': 'P0C6U8', 'xref_name': None, 'xre...   
7                                                 []   
8  [{'xref_id': 'P0C6X7', 'xref_name': None, 'xre...   
9                                                 []   

                                            organism  \
0                                        Coronavirus   
1                                 Feline coronavirus   
2                                 Murine coronavirus   
3                                 Canine coronavirus   
4                             Human coronavirus 229E   
5                             Human coronavirus OC43   
6                                   SARS coronavirus   
7  Middle East respiratory syndrome-related coron...   
8                                   SARS coronavirus   
9    Severe acute respiratory syndrome coronavirus 2   

                                           pref_name  score  \
0                                        Coronavirus   17.0   
1                                 Feline coronavirus   14.0   
2                                 Murine coronavirus   14.0   
3                                 Canine coronavirus   14.0   
4                             Human coronavirus 229E   13.0   
5                             Human coronavirus OC43   13.0   
6                SARS coronavirus 3C-like proteinase   10.0   
7  Middle East respiratory syndrome-related coron...    9.0   
8                          Replicase polyprotein 1ab    4.0   
9                          Replicase polyprotein 1ab    4.0   

   species_group_flag target_chembl_id  \
0               False     CHEMBL613732   
1               False     CHEMBL612744   
2               False    CHEMBL5209664   
3               False    CHEMBL5291668   
4               False     CHEMBL613837   
5               False    CHEMBL5209665   
6               False       CHEMBL3927   
7               False    CHEMBL4296578   
8               False       CHEMBL5118   
9               False    CHEMBL4523582   

                                   target_components     target_type   tax_id  
0                                                 []        ORGANISM    11119  
1                                                 []        ORGANISM    12663  
2                                                 []        ORGANISM   694005  
3                                                 []        ORGANISM    11153  
4                                                 []        ORGANISM    11137  
5                                                 []        ORGANISM    31631  
6  [{'accession': 'P0C6U8', 'component_descriptio...  SINGLE PROTEIN   227859  
7                                                 []        ORGANISM  1335626  
8  [{'accession': 'P0C6X7', 'component_descriptio...  SINGLE PROTEIN   227859  
9  [{'accession': 'P0DTD1', 'component_descriptio...  SINGLE PROTEIN  2697049

### 2.3.2 Select and Retrieve Data for SARS Coronavirus 3C-like Proteinase (sevent entry)

Seventh entry (which corresponds to the target protein, coronavirus 3C-like proteinase) to the **selected_target** variable

In [5]:
# save target chembl id into selected_target variable
selected_target = targets.target_chembl_id[6]
selected_target

'CHEMBL3927'

Here, we will retrieve only bioactivity data for *coronavirus 3C-like proteinase* (CHEMBL3972) that are reported as $IC_{50}$ values in nM(nanomolar) unit.

In [6]:
# retrieve bioactivity data and save it to res variable
activity = new_client.activity
res = activity.filter(target_chembl_id=selected_target).filter(standard_type="IC50")

In [7]:
# save res data into dataframe
df = pd.DataFrame.from_dict(res)
df

action_type activity_comment  activity_id activity_properties  \
0          None             None      1480935                  []   
1          None             None      1480936                  []   
2          None             None      1481061                  []   
3          None             None      1481065                  []   
4          None             None      1481066                  []   
..          ...              ...          ...                 ...   
128        None             None     12041507                  []   
129        None             None     12041508                  []   
130        None             None     12041509                  []   
131        None             None     12041510                  []   
132        None             None     12041511                  []   

    assay_chembl_id                                  assay_description  \
0      CHEMBL829584  In vitro inhibitory concentration against SARS...   
1      CHEMBL829584  In vitro inhibitory concentration against SARS...   
2      CHEMBL830868  In vitro inhibitory concentration against SARS...   
3      CHEMBL829584  In vitro inhibitory concentration against SARS...   
4      CHEMBL829584  In vitro inhibitory concentration against SARS...   
..              ...                                                ...   
128   CHEMBL2150313  Inhibition of SARS-CoV PLpro expressed in Esch...   
129   CHEMBL2150313  Inhibition of SARS-CoV PLpro expressed in Esch...   
130   CHEMBL2150313  Inhibition of SARS-CoV PLpro expressed in Esch...   
131   CHEMBL2150313  Inhibition of SARS-CoV PLpro expressed in Esch...   
132   CHEMBL2150313  Inhibition of SARS-CoV PLpro expressed in Esch...   

    assay_type assay_variant_accession assay_variant_mutation bao_endpoint  \
0            B                    None                   None  BAO_0000190   
1            B                    None                   None  BAO_0000190   
2            B                    None                   None  BAO_0000190   
3            B                    None                   None  BAO_0000190   
4            B                    None                   None  BAO_0000190   
..         ...                     ...                    ...          ...   
128          B                    None                   None  BAO_0000190   
129          B                    None                   None  BAO_0000190   
130          B                    None                   None  BAO_0000190   
131          B                    None                   None  BAO_0000190   
132          B                    None                   None  BAO_0000190   

     ...   target_organism                     target_pref_name target_tax_id  \
0    ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
1    ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
2    ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
3    ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
4    ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
..   ...               ...                                  ...           ...   
128  ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
129  ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
130  ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
131  ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
132  ...  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   

    text_value  toid  type units    uo_units upper_value  value  
0         None  None  IC50    uM  UO_0000065        None    7.2  
1         None  None  IC50    uM  UO_0000065        None    9.4  
2         None  None  IC50    uM  UO_0000065        None   13.5  
3         None  None  IC50    uM  UO_0000065        None  13.11  
4         None  None  IC50    uM  UO_0000065  

In [8]:
# check data variant on standard_type column
df.standard_type.unique()

array(['IC50'], dtype=object)

Finally, resulting bioactivity data into CSV file **bioactivity_data.csv**

In [9]:
df.to_csv('bioactivity_data.csv', index=False)

## 2.4 Data Overview

In [10]:
# show sample data
df.head(5)

action_type activity_comment  activity_id activity_properties  \
0        None             None      1480935                  []   
1        None             None      1480936                  []   
2        None             None      1481061                  []   
3        None             None      1481065                  []   
4        None             None      1481066                  []   

  assay_chembl_id                                  assay_description  \
0    CHEMBL829584  In vitro inhibitory concentration against SARS...   
1    CHEMBL829584  In vitro inhibitory concentration against SARS...   
2    CHEMBL830868  In vitro inhibitory concentration against SARS...   
3    CHEMBL829584  In vitro inhibitory concentration against SARS...   
4    CHEMBL829584  In vitro inhibitory concentration against SARS...   

  assay_type assay_variant_accession assay_variant_mutation bao_endpoint  ...  \
0          B                    None                   None  BAO_0000190  ...   
1          B                    None                   None  BAO_0000190  ...   
2          B                    None                   None  BAO_0000190  ...   
3          B                    None                   None  BAO_0000190  ...   
4          B                    None                   None  BAO_0000190  ...   

    target_organism                     target_pref_name target_tax_id  \
0  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
1  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
2  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
3  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   
4  SARS coronavirus  SARS coronavirus 3C-like proteinase        227859   

  text_value  toid  type units    uo_units upper_value  value  
0       None  None  IC50    uM  UO_0000065        None    7.2  
1       None  None  IC50    uM  UO_0000065        None    9.4  
2       None  None  IC50    uM  UO_0000065        None   13.5  
3       None  None  IC50    uM  UO_0000065        None  13.11  
4       None  None  IC50    uM  UO_0000065        None    2.0  

[5 rows x 46 columns]

In [11]:
# show dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 133 entries, 0 to 132
Data columns (total 46 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   action_type                0 non-null      object
 1   activity_comment           0 non-null      object
 2   activity_id                133 non-null    int64 
 3   activity_properties        133 non-null    object
 4   assay_chembl_id            133 non-null    object
 5   assay_description          133 non-null    object
 6   assay_type                 133 non-null    object
 7   assay_variant_accession    0 non-null      object
 8   assay_variant_mutation     0 non-null      object
 9   bao_endpoint               133 non-null    object
 10  bao_format                 133 non-null    object
 11  bao_label                  133 non-null    object
 12  canonical_smiles           133 non-null    object
 13  data_validity_comment      27 non-null     object
 14  data_valid

**Findings**
1. Based on the information above, it is know that dataset `df` has 46 columns with 133 entries.
2. There are some columns with different data types, 6 columns with integer type and 40 columns with object type.

In [12]:
# check missing values on standard_value column
df.standard_value.isna().sum()

np.int64(0)

There is no missing value for `standard_value` column.

## 2.5 Data Pre-Processing of the Bioactivity Data

### 4.1 Labelling Compounds as either being Active, Inactive, or Intermediate

The bioactivity data is in the $IC_{50}$ unit. Compounds having standard values of less than 1,000 nM will be considered to be **active** while those greater than 10,000 nM will be **inactive**. As for those values in between 1,000 and 10,000 nM will be referred to as **intermediate**.

In [13]:
# labelling compouds based on standard_value
# save new label into bioactivity_class variable
bioactivity_class = []
for i in df.standard_value:
    if float(i) >= 10000:
        bioactivity_class.append('inactive')
    elif float(i) <= 1000:
        bioactivity_class.append('active')
    else:
        bioactivity_class.append('intermediate')

## 4.1 Retrive necessary information from `df` dataset

In [17]:
# gather neccesary information from df dataset
selection = ['molecule_chembl_id', 'canonical_smiles', 'standard_value']
df2 = df[selection]

### Combine `Biocativity_class` lists above and df2 dataset

In [25]:
df_temp = pd.concat([df2, pd.Series(bioactivity_class)], axis=1)
df_temp

molecule_chembl_id                                   canonical_smiles  \
0         CHEMBL187579             Cc1noc(C)c1CN1C(=O)C(=O)c2cc(C#N)ccc21   
1         CHEMBL188487             O=C1C(=O)N(Cc2ccc(F)cc2Cl)c2ccc(I)cc21   
2         CHEMBL185698            O=C1C(=O)N(CC2COc3ccccc3O2)c2ccc(I)cc21   
3         CHEMBL426082                O=C1C(=O)N(Cc2cc3ccccc3s2)c2ccccc21   
4         CHEMBL187717    O=C1C(=O)N(Cc2cc3ccccc3s2)c2c1cccc2[N+](=O)[O-]   
..                 ...                                                ...   
128      CHEMBL2146517  COC(=O)[C@@]1(C)CCCc2c1ccc1c2C(=O)C(=O)c2c(C)c...   
129       CHEMBL187460        C[C@H]1COC2=C1C(=O)C(=O)c1c2ccc2c1CCCC2(C)C   
130       CHEMBL363535             Cc1coc2c1C(=O)C(=O)c1c-2ccc2c(C)cccc12   
131       CHEMBL227075      Cc1cccc2c3c(ccc12)C1=C(C(=O)C3=O)[C@@H](C)CO1   
132        CHEMBL45830             CC(C)C1=Cc2ccc3c(c2C(=O)C1=O)CCCC3(C)C   

    standard_value             0  
0           7200.0  intermediate  
1           9400.0  intermediate  
2          13500.0      inactive  
3          13110.0      inactive  
4           2000.0  intermediate  
..             ...           ...  
128        10600.0      inactive  
129        10100.0      inactive  
130        11500.0      inactive  
131        10700.0      inactive  
132        78900.0      inactive  

[133 rows x 4 columns]

In [27]:
# rename the column
df3 = df_temp.rename(columns={0: 'bioactivity_data'})
df3

molecule_chembl_id                                   canonical_smiles  \
0         CHEMBL187579             Cc1noc(C)c1CN1C(=O)C(=O)c2cc(C#N)ccc21   
1         CHEMBL188487             O=C1C(=O)N(Cc2ccc(F)cc2Cl)c2ccc(I)cc21   
2         CHEMBL185698            O=C1C(=O)N(CC2COc3ccccc3O2)c2ccc(I)cc21   
3         CHEMBL426082                O=C1C(=O)N(Cc2cc3ccccc3s2)c2ccccc21   
4         CHEMBL187717    O=C1C(=O)N(Cc2cc3ccccc3s2)c2c1cccc2[N+](=O)[O-]   
..                 ...                                                ...   
128      CHEMBL2146517  COC(=O)[C@@]1(C)CCCc2c1ccc1c2C(=O)C(=O)c2c(C)c...   
129       CHEMBL187460        C[C@H]1COC2=C1C(=O)C(=O)c1c2ccc2c1CCCC2(C)C   
130       CHEMBL363535             Cc1coc2c1C(=O)C(=O)c1c-2ccc2c(C)cccc12   
131       CHEMBL227075      Cc1cccc2c3c(ccc12)C1=C(C(=O)C3=O)[C@@H](C)CO1   
132        CHEMBL45830             CC(C)C1=Cc2ccc3c(c2C(=O)C1=O)CCCC3(C)C   

    standard_value bioactivity_data  
0           7200.0     intermediate  
1           9400.0     intermediate  
2          13500.0         inactive  
3          13110.0         inactive  
4           2000.0     intermediate  
..             ...              ...  
128        10600.0         inactive  
129        10100.0         inactive  
130        11500.0         inactive  
131        10700.0         inactive  
132        78900.0         inactive  

[133 rows x 4 columns]

In [29]:
df3.to_csv('bioactivity_preprocessed_data.csv', index=False)